lib/navigation/preprocess_imu.cpp

```diff
@@ -1,56 +1,137 @@
 #include "preprocess_imu.hpp"
 
+#include <cmath>
+
+#include <algorithm>
+#include <numeric>
+#include <optional>
+
 namespace hyped::navigation {
 
-ImuPreprocessor::ImuPreprocessor()
+ImuPreprocessor::ImuPreprocessor(core::ILogger &logger)
+    : log_(logger),
+      num_outliers_per_imu_({0, 0, 0, 0}),
+      are_imus_reliable_({true, true, true, true}),
+      num_reliable_accelerometers_(core::kNumImus)
 {
-  // TODOLater: implement
 }
 
-core::ImuData ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
+std::optional<core::ImuData> ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkReliable
-
-  */
-  return {0, 0, 0, 0};
+  core::ImuData imu_data;
+  // TODOLater : check on direction of travel
+  core::Float magnitude;
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    magnitude = 0;
+    for (std::size_t j = 0; j < 3; ++j) {
+      magnitude += std::pow(raw_imu_data.at(i).at(j), 2);
+    }
+    imu_data.at(i) = std::sqrt(magnitude);
+  }
+
+  const core::ImuData accelerometer_data = detectOutliers(imu_data);
+  checkReliable();
+  return accelerometer_data;
+}
+
+core::ImuData ImuPreprocessor::detectOutliers(const core::ImuData imu_data)
+{
+  core::ImuData accelerometer_data;
+  for (std::size_t i; i < core::kNumImus; ++i) {
+    accelerometer_data.at(i) = imu_data.at(i);
+  }
+
+  Quartiles quartiles;
+  if (num_reliable_accelerometers_ == core::kNumImus) {
+    quartiles = getQuartiles(imu_data);
+  } else {
+    std::array<core::Float, core::kNumImus - 1> faulty_data;
+    std::size_t index;
+    for (std::size_t i = 0; i < core::kNumImus - 1; ++i) {
+      if (!are_imus_reliable_.at(i)) { index = i; }
```

Comment:
Here we are making the strong assumption that there is only one faulty IMU. This will be a hassle in case it ever changes and ... we don't really need to?

Why not 
```cpp
std::size_t j = 0;
for (std::size_t i = 0; i < core::kNumImus; ++i) {
  if (are_imus_reliable_.at(i)) {
    filtered_data.at(j) = imu_data.at(i);
    ++j;
  }
 }
```

Of course, you now have to do some thinking about how to make things work for arbitrary sets of data, dependent on the values provided at runtime. An obvious choice is `std::vector` but we need to make sure not to allocate. What we really want is a collection that behaves like an `std::vector` except that we guarantee that we are not going to add more than `kNumImus` elements, ever. This allows for an underlying structure `std::array<T, kNumImus>`. I guess the simplest way of doing this is to simply use `std::array` everywhere and pass an additional `std::size_t length` like in the good old C days? @kshxtij

Reply:
You could then add a `// TODOLater` to make it into a nice data structure... which I might eventually do.